### ISSUE TITLE
TileSize 32 and performance of AMD GCN

### ISSUE THREAD

**Issue by ThWuensche** [openmm/openmm#2875]
This is more a question than an issue. I'm trying to understand why AMD GCN (Radeon VII) on many projects provides low throughput. Browsing through the code I find TileSize set to 32, while GCN has 64 parallel threads in a CU.

Does the tile size of 32 still use the full width of the GCN CUs? Or is the TileSize completely unrelated to simdWidth, which is 64?

Sorry for asking beginners questions, if too stupid, please just do not respond.

**Comment by peastman** [comment]
It's mostly, but not entirely unrelated.  The performance is best when the tile size matches the SIMD width.  If the tile size is larger than the SIMD width, we have to add extra synchronization calls which can impact performance.  We do that on Intel GPUs which have a width of only 16.  If the tile size is smaller than the SIMD width everything mostly behaves as normal, but you get a little bit more thread divergence which hurts performance a little bit.  That's the case on GCN.  In AMD's latest generation of GPUs they reduced the width to 32, so that isn't an issue anymore.

If you look through the OpenCL kernels (for example, nonbonded.cl), you'll find some lines that say `SYNC_WARPS`.  That symbol gets defined based on the width:

https://github.com/openmm/openmm/blob/94d7225bfe59e2ce80fc4edfc1828d15d73aa1f5/platforms/opencl/src/OpenCLContext.cpp#L283-L286

If the width is less than 32 it gets defined as a barrier to perform synchronization.  Otherwise, it gets defined as a much less expensive memory fence that just ensures memory access gets ordered properly without requiring threads to block.

**Comment by ThWuensche** [comment]
Thanks for your answer. I was in fear that a tile size at 1/2 of simdWidth might leave 1/2 of the possible simd threads in a CU inactive/idle, thus reducing performance of the step to 1/2 of the possible performance.

For information: I had tried to increase tile size to 64 on my GCN device, but that resulted in NaN very early. So the code seems to depend on something else, does not allow simple change of tile size.

**Comment by peastman** [comment]
In what way did you try to run it on two GPUs in parallel?  By setting the `DeviceIndex` property to a list of two GPUs?  Or running two simulations in different processes?  Or is this based on some sort of code change?

**Comment by ThWuensche** [comment]
I had changed my test script from the performance tests to include:

platform.setPropertyDefaultValue('DeviceIndex', '0,1')

**Comment by peastman** [comment]
That ought to work fine.  What OpenCL implementation did you use?

**Comment by ThWuensche** [comment]
ROCm 3.7.0

**Comment by ThWuensche** [comment]
INVCUT6 seems to be created in OpenCLKernels, but only if pmeio == NULL. pmeio is set above if useCpuPME.

```
            if (pmeio == NULL) {
                // Create required data structures.

                if (doLJPME) {
                    double invRCut6 = pow(force.getCutoffDistance(), -6);
                    double dalphaR = dispersionAlpha * force.getCutoffDistance();
                    double dar2 = dalphaR*dalphaR;
                    double dar4 = dar2*dar2;
                    double multShift6 = -invRCut6*(1.0 - exp(-dar2) * (1.0 + dar2 + 0.5*dar4));
                    defines["INVCUT6"] = cl.doubleToString(invRCut6);
                    defines["MULTSHIFT6"] = cl.doubleToString(multShift6);
                }



```

**Comment by ThWuensche** [comment]
Ok, it seems to depend on the simulated system. The error occured on RUN0 from the F@H benchmarks, RUN10 and RUN11 seem to work. However on RUN10 the ns/day from 2 GPUs came out the same as for one GPU :(

**Comment by ThWuensche** [comment]
This is statistics result from F@H benchmark unit RUN9:
![statistics_run9](https://user-images.githubusercontent.com/70173377/94993488-b7fadd00-0591-11eb-9a67-33dcecfd7ae2.png)
It shows extremely low performance, also in comparision to "wide" (high compute unit count) nVidia GPUs. That probably can not be explained purely by hardware characteristics. One reason might be kernel startup time/overhead much higher than that for nVidia GPUs, which would have high effect specially in short kernel runs. Another one might be implementation details in the openMM kernel code, which are different for the two GPU manufacturers. Such implementation details exist based on simdWidth, but probably have been placed to improve efficiency. If you have any hints what to look for, it would be welcome.

**Comment by peastman** [comment]
I'm not clear on what the graph is showing.  What's the difference between "OpenCL" and "openMMOCL"?  Which of those, if any, is using two GPUs?

**Comment by ThWuensche** [comment]
All of them are using only one GPU, considering the original intent of my post about finding ways to improve the performance on AMD GCN. The data for Radeon VII has been added to the diagram created by PantherX / F@H. Column OpenCL is showing the values from F@H benchmark runs for the nVidia GPUs, for AMD Radeon VII the values from the F@H logs on my machine (divided by 4 to compensate for the bug in calculating that value). Column openMMHIP holds figures for the HIP build of openMM executed through the python test script, openMMOCL is the same for platform openCL. All of that shows the extremely low performance of Radeon VII (gfx906) on the small system ~4000 atoms compared to nVidia. For large systems the performance is much better / more competitive.

**Comment by ThWuensche** [comment]
I'm trying my luck in tracing/profiling. The following shows that a very big part of time is spent in sorting:
![run9_trace](https://user-images.githubusercontent.com/70173377/94997129-868e0b80-05a9-11eb-8143-f1d0cddfe227.png)

**Comment by peastman** [comment]
That's very strange.  That kernel ought to be only a minor contribution to the time.  Is that the OpenCL or HIP version?

**Comment by ThWuensche** [comment]
It's the openCL version, with a small experimental change in OpenCLNonbondedUtilities.cpp:

```
//    sort(exclusionTilesVec.begin(), exclusionTilesVec.end(), context.getSIMDWidth() <= 32 ? compareInt2 : compareInt2LargeSIMD);
    sort(exclusionTilesVec.begin(), exclusionTilesVec.end(), compareInt2);

```
Will revert that change and retry, but it did not have effect on overall performance

**Comment by ThWuensche** [comment]
Reverting the change does not change the situation. I have to say that I'm only in the process of getting familiar with profiling tools and did not yet find a good solution. The data has been taken with ROCm rocprof, display is with catapult tool (github.com/catapult-project), since codeXL isn't supported any more, so do not see a reasonable toolchain from AMD for that in the moment. But still I hope that the information from the trace is correct.

Here is the browser window in total, the part in the lower part is reported after selecting a range on the light red part in the timeline GPU in the top:
![grafik](https://user-images.githubusercontent.com/70173377/94998298-2e5b0780-05b1-11eb-8651-b7d7a07e4a13.png)

**Comment by ThWuensche** [comment]
OK, I just commented out the whole sortShortList part and performance improves from 89.5ns/day to 132.4ns/day for benchmark RUN9

**Comment by peastman** [comment]
Try modifying this line to force `isShortList` to be false:

https://github.com/openmm/openmm/blob/94d7225bfe59e2ce80fc4edfc1828d15d73aa1f5/platforms/opencl/src/OpenCLSort.cpp#L73

That will make it always use the standard sorting method, which involves multiple kernels.  How does that affect performance?

Another thing you could try is modifying the lines just below that to set `useShortList2` to true (with `isShortList` also true).  That uses an alternate version of the `sortShortList()` kernel that we normally only use on NVIDIA.

**Comment by ThWuensche** [comment]
> Try modifying this line to force `isShortList` to be false:
> 
> https://github.com/openmm/openmm/blob/94d7225bfe59e2ce80fc4edfc1828d15d73aa1f5/platforms/opencl/src/OpenCLSort.cpp#L73
> 
> That will make it always use the standard sorting method, which involves multiple kernels. How does that affect performance?

That probably results in the same as commenting out that part.

> Another thing you could try is modifying the lines just below that to set `useShortList2` to true (with `isShortList` also true). That uses an alternate version of the `sortShortList()` kernel that we normally only use on NVIDIA.

Just tried that on RUN16, result is 47ns/day, so on the same level as the other implementation and slower than the run without any sortShortList. Will try RUN9 now.

**Comment by ThWuensche** [comment]
I don't know whether it helps, but here is the trace file saved for 500 steps of RUN9. As mentioned the interesting thing is the large difference, from 2 runs of sortShortList one takes about 13us, the next about 400us.
[results.json.zip](https://github.com/openmm/openmm/files/5322234/results.json.zip)

**Comment by peastman** [comment]
That sounds like it's using the short list kernel even when it isn't appropriate.  At the point where it sets `isShortList`, try adding this line:

```c++
printf("%d %d %d %d %d %d\n", length, maxShortList, maxLocalBuffer, context.getNumThreadBlocks(), maxSharedMem, trait->getDataSize());
```

That should show us what it's doing and why.

**Comment by ThWuensche** [comment]
That's the output:

4071 8192 4096 960 65536 8
128 8192 4096 960 65536 8

**Comment by peastman** [comment]
In the line

https://github.com/openmm/openmm/blob/94d7225bfe59e2ce80fc4edfc1828d15d73aa1f5/platforms/opencl/src/OpenCLSort.cpp#L66

try changing 8192 to 1024.  That should make it continue to use the short list kernel for the small list (length 128) but the standard kernels for the longer one (length 4071).  How is the speed?  Can you post the profiler output for that case?

**Comment by ThWuensche** [comment]
Delivers 134.8ns/day for RUN9, much better than before and slightly more than for the general sort. Will now generate profiler output.

**Comment by ThWuensche** [comment]
Here is the output of rocprof:
[results.json.zip](https://github.com/openmm/openmm/files/5322273/results.json.zip)

**Comment by peastman** [comment]
That all looks reasonable.  So the question is, why is `sortShortList()` so slow on this GPU for a list with only 4071 elements?  It easily fits in shared memory, and it ought to be very fast.  Can the profiler give any more details about the bottlenecks in the kernel?

Unfortunately I don't currently have access to a GCN GPU.  I do have a Navi card, and I'll see if I can do some tests on it in the next couple of days.

**Comment by ThWuensche** [comment]
I will try to find out, should learn more about profiling GPU kernels. If you want me to test something on GCN, please let me know.

It's late here already in Germany, so that will be for another day. Thanks for your support, sorry, that I can't improve things myself so far, but hope looking out for non-reasonable things and asking still is of help.

**Comment by peastman** [comment]
What you're doing is fantastic.  We've already narrowed it down to a particular kernel that's slow in a particular situation, which is excellent progress.  If we can't track it any further, we could just decrease the limit on the list size to use that kernel on AMD GPUs.  But if possible, I'd like to understand better what's going on, since there might be a better solution.

**Comment by ThWuensche** [comment]
Just for me to understand: That sort is executed in exactly one compute unit (out of the 60 that are on the chip in case of Radeon VII)? The device has LDS (local data share, individual to CU), GDS (global data share, available to all CUs, on-chip) and device memory (off-chip, slower). Where is input data residing, first in device memory, then in the kernel copied to LDS? Or is it sitting initially (at kernel start) in GDS?

I don't understand the parameter lists.

For the kernel:
__kernel void sortShortList(__global DATA_TYPE* restrict data, uint length, __local DATA_TYPE* dataBuffer) 

In the call:
                shortListKernel.setArg<cl::Buffer>(0, data.getDeviceBuffer());
                shortListKernel.setArg<cl_uint>(1, dataLength);
                shortListKernel.setArg(2, dataLength*trait->getDataSize(), NULL);
                context.executeKernel(shortListKernel, sortKernelSize, sortKernelSize);

The call looks as if param1 would be the pointer in device memory, param2 the number of elements, param3 the number of bytes. How does that correlate with the kernel parameter list?  Arg2 is a size, however the kernel expects a memory location in local memory? Might that result in an invalid memory address for what is expected to be a local memory buffer?

**Comment by informatorius** [comment]
Maybe you want to try that too for AMD
https://github.com/openmm/openmm/issues/2782

**Comment by ThWuensche** [comment]
I tried creating the buffer in local memory directly in the kernel, seems to work, but does not give improved performance:

`__local DATA_TYPE dataBuffer[4096];
`

Or would that create individual buffers in each thread of a CU?

What also makes me wonder, whether each thread in a CU would copy the memory, though LDS is shared for all the threads in that CU, thus creating additional memory transfer operations to fill common memory:

```
for (int index = get_local_id(0); index < length; index += get_local_size(0))
    dataBuffer[index] = data[index];
```

Just arbitrary ideas, experimenting.

**Comment by ThWuensche** [comment]
@informatorius Thanks, will have a look at that.

**Comment by ThWuensche** [comment]
In the meantime I wonder whether for large data sets the standard algorithm, which runs as far as I understand on many compute units in parallel, is not simply more efficient than the sortShortList algorithm, which can use local memory, but as far as I understand is executed on only one compute unit.

As an experiment I have guarded the memcopy loops with:

`if (get_global_id(0) < get_local_size(0))`

as I guessed on lengths > local_size the copy could be done repeatedly, not sure whether that is correct, but seems to me.

I've tried to run sortShortList directly on global memory and that increases the time of the sort by about 60%, so memory bandwidth is an issue, but maybe not as serious as limiting search to one CU instead of distribution to many CUs with the standard algorithm. So so far my understanding is that the switch between sortShortList and standard sort might be set to a smaller size of the data to be searched.

**Comment by peastman** [comment]
The arguments to the kernel are

- A pointer to device memory containing the data to sort
- The length of the list to sort
- A local memory buffer for use as workspace

OpenCL has two ways to create a local memory buffer.  If it always has the same length, you can declare it directly the way you did:

```c++
__local DATA_TYPE dataBuffer[4096]; 
```

But in this case, the size of the buffer depends on the length of the list to sort.  In that case, you instead include it in the argument list with the `__local` qualifier.  The host can then use `setArg()` to set the length of it:

```c++
shortListKernel.setArg(2, dataLength*trait->getDataSize(), NULL);
```

> What also makes me wonder, whether each thread in a CU would copy the memory, though LDS is shared for all the threads in that CU, thus creating additional memory transfer operations to fill common memory:

The threads are all working together to load the data.  Notice the loop starts out `index = get_local_id(0)`.  Every thread is loading a different element.  Then they all step forward by the number of threads in the group: `index += get_local_size(0)`.  Each element gets loaded once by one thread.

**Comment by ThWuensche** [comment]
Thanks, after some hours of studying that's what I also understood. One question is whether for length > local_size the copy is repeated, since local_id will start from 0 again. But even if that would be the case in the end this seems to have very little effect on throughput. Learned today that a printf in kernel code works, so feeling much more "at home" now :). With the change of limit to decide between short and standard to 1024 elements on the example the performance raised from 89.5ns/day to 137.5ns/day, so at least something achieved. Will continue by further studying the profiler trace.

**Comment by ThWuensche** [comment]
When I build with TileSize 64 I get NaNs. Is that expected behaviour? If not, what should I look for to fix it? I already have searched for "32", assuming that somewhere a dependancy is hard-coded. There are some occurences, but hard to say whether or what is linked to the problem.

**Comment by peastman** [comment]
There's actually three different sorting algorithms.  The standard method uses a bucket sort, followed by a bitonic merge sort in each bucket.  That's a very efficient algorithm, but it requires five kernels to execute the whole thing.  Each kernel has overhead, and when you're sorting a short list that overhead ends up dominating.  The overhead from each kernel is larger than the time spent actually executing it.

`sortShortList()` copies the entire list to local memory, then uses a bitonic merge sort.  That has to be done in a single workgroup, so most of the GPU sits idle.  But you only have to launch one kernel, so for short lists it ends up being faster.

`sortShortList2()` uses a totally different algorithm.  Each thread picks a single element of the list and does a full scan through the list to figure out where it should be stored.  This is intrinsically a less efficient algorithm, but it also parallelizes perfectly.  I benchmarked it and found that it was faster on NVIDIA GPUs, but slower on AMD GPUs, so I set it to be used only on NVIDIA.

Did you try setting `useShortList2` to true?

> When I build with TileSize 64 I get NaNs. Is that expected behaviour?

Increasing the tile size to larger than 32 would take major changes.  For example, we use 32 bit ints as bit flags for the atoms in a block at various points, so all of those would need to be changed.

**Comment by ThWuensche** [comment]
Thanks for the explanation of the different algorithms. So the question is when to switch over from the sort algorithms for short lists to the one for long lists. So far it is done at maximum at the level of 8192 elements, that limit is set by the level of available local memory and threadBlockSize*numThreadBlocks. But from what I have seen for optimum performance the limit at least for sortShortList should be much lower, since that algorithm delivers low performance on larger data due to the "one CU" implementation. For sortShortList2 it may be different, since there the limit is the available threads and memory as implemented in the decision in line 66 of OpenCLSort.cpp.

I had once tried sortShortList2, the results (overall from the project run) looked similar to those of sortShortList, but that was only a very short try. Consider to reevaluate that.

Unfortunately in the moment I have a test case only for 128 elements and slightly beyond 4000 elements, where the sortShortList implementation for the 4000 elements obviously is very slow. For benchmarking it would be good to have finer grained levels for decision, like 128, 125, 512, 1024, 2048, 4096, 8192 elements. Benchmarking could be done either by a separate test setup not running a real simulation, but calling directly the sort routines. However another and maybe easier approach would be to set up a simulation which just includes blocks to sort of that sizes and look on the overall results and maybe traces from a profiler. How difficult would it be to set up such system with the tools you or John Chodera have at hand? I think such setup is done frequently, so hopefully not big effort.

As for the larger tile size I just wanted to try the effect, but as it needs significant modification that's obviously not worth the effort.

**Comment by peastman** [comment]
> So far it is done at maximum at the level of 8192 elements, that limit is set by the level of available local memory and threadBlockSize*numThreadBlocks.

The upper limit of 8192 was set by testing on a variety of GPUs, both NVIDIA and AMD, to see where the performance crossover happened.  So the question is, why is that routine so slow on your GPU?

> Benchmarking could be done either by a separate test setup not running a real simulation, but calling directly the sort routines.

Take a look at TestOpenCLSort.cpp.

**Comment by ThWuensche** [comment]
> > So far it is done at maximum at the level of 8192 elements, that limit is set by the level of available local memory and threadBlockSize*numThreadBlocks.
> 
> The upper limit of 8192 was set by testing on a variety of GPUs, both NVIDIA and AMD, to see where the performance crossover happened. So the question is, why is that routine so slow on your GPU?

Good question, no idea. There might be a problem in the fact that local memory is organized in banks, don't know about timing in parallel access case.

> > Benchmarking could be done either by a separate test setup not running a real simulation, but calling directly the sort routines.
> 
> Take a look at TestOpenCLSort.cpp.

Thanks, just should open my eyes :(

By default the test script would execute on the CPU, as it seems (device 0 is CPU, at least in ROCm). I have modified TestOpenCLSort to run first the two long runs as in the original code and then the short variants for 128, 256, 512, 1024, 2048, 4096, 8192 list elements. Five runs of each variant have been executed. Here are the values extracted from the profiler logs. Only on rather small lists a benefit from the sortShortList variants is visible. Sort1Testx is from sortShortList, Sort2Testx from sortShortList2, sortLTestx with isShortList set to false.

[sort_timing.txt](https://github.com/openmm/openmm/files/5330111/sort_timing.txt)

If I run on CPU and time the call, I get constant and very short times. Guess, the sorting is done not in the main thread, but in a parallel thread and the sort call returns before the sort is done, Is that correct?

**Comment by peastman** [comment]
Here are the times I get for `sortShortList()` running on a Radeon RX 5700 XT.

| Length | Time (us) |
| ------ | --------- |
| 100 | 25 |
| 500 | 95 |
| 1000 | 148 |
| 1500 | 240 |
| 2000| 301 |
| 2500|437|
|3000|513|
|3500|592|
|4000|668|

That's really slow.  In contrast, using the standard sort algorithm the total time for the five kernels increases from 49.7 at length 100 to 77 at length 4000.

**Comment by ThWuensche** [comment]
Can not generate such nice table:
```
Size | short | short2 | std
128 | 18 | 22 | 36
256 | 22 | 39 | 43
512 | 41 | 71 | 91
1024 | 86 | 137 | 131
2048 | 196 | 268 | 163
4096 | 480 | 530 | 167
8192 | 1101 | 1144 | 236
Kernel execution times in us


```

**Comment by ThWuensche** [comment]
So compared to the 5700XT sortShortList is faster on Radeon VII, times for standard also seem to be somewhat faster.

**Comment by ThWuensche** [comment]
Break even would be somewhere around 1000 elements, as it seems.

**Comment by ThWuensche** [comment]
I don't have any NVidia GPU to compare.

**Comment by jchodera** [comment]
Could an adaptive scheme automatically select whichever algorithm is fastest for the given system so we don't hard-code this for specific GPUs?

**Comment by ThWuensche** [comment]
> That's really slow. In contrast, using the standard sort algorithm the total time for the five kernels increases from 49.7 at length 100 to 77 at length 4000.

Is that 77 at length 4000 for log distribution or even distribution? That seems to make a difference and my measurements are based on log distribution as in TestOpenCLSort.

**Comment by peastman** [comment]
Here are the times for a Titan V.

|Length|Time|
|---|---|
|500|25|
|1000|52|
|1500|88|
|2000|112|
|2500|165|
|3000|195|
|3500|227|
|4000|256|

But of course on NVIDIA GPU's we use `sortShortList2()` instead.  On AMD it's consistently a bit slower than `sortShortList()`.  Here are the times for it on Titan V.

|Length|Time|
|---|---|
|100|13|
|500|18|
|1000|26|
|1500|36|
|2000|47|
|2500|56|
|3000|65|
|3500|75|
|4000|84|

Finally I looked at the time for the standard sort kernels.  They barely vary at all, going from 34 us for 100 elements to 39 us for 4000 elements.

Here are the conclusions I draw.

- On NVIDIA `sortShortList2()` is faster than `sortShortList()` in all cases.  On AMD it's slower in all cases.
- On AMD we should set the limit for the short list kernel very low.  It's faster than the standard algorithm at 100 elements, but by 500 it's already slower.
- On NVIDIA the limit should be higher.  On Titan V the crossover seems to happen somewhere around 1500 elements.

**Comment by ThWuensche** [comment]
For the standard sort 10000 elements (as in TestOpenCLSort) I have 67us for even distribution, 227 for log distribution. Might need a test for even distribution on lower element counts.

**Comment by peastman** [comment]
The distribution doesn't affect the short list algorithms.  It does affect the standard algorithm, since the bucket sort isn't as effective if the distribution isn't uniform.  All of the above tests were for uniform.

**Comment by ThWuensche** [comment]
Ok, that explains the 77 for 4000 standard on the 5700XT compared to my results on the Radeon VII with log distribution.

Anyhow I'm off to bed now, it's 1:34 in the night/morning here.

**Comment by peastman** [comment]
Of course, microbenchmarks don't always reflect real world performance.  The ApoA1 benchmark involves sorting a list of length 2882.  That's well above the crossover point identified above, so switching it to the standard algorithm instead of the short list algorithm should make the simulation faster, right?  But in fact it makes it slower.  By a mostly negligible amount for OpenCL, but almost 5% for CUDA.

What's going on here...

**Comment by peastman** [comment]
Based on the CUDA profile, I *think* this is what's happening.

We split the computation between two streams, which allows the GPU to execute two kernels at once.  It turns out that sorting the list is done on one stream at the same time that `gridSpreadCharge()` is running on the other stream.  That's a somewhat expensive kernel, and it's performance is entirely limited by memory access.  It does a ton of writes to global memory.  The short list kernel takes longer to run than the standard sorting algorithm, but it also involves a lot less memory access.  That means it leaves the bus free for the charge spreading kernel, which ends up making the simulation faster.

**Comment by ThWuensche** [comment]
As mentioned above, for the RUN9 of F@H benchmarks the limit at 1024 improves the estimated ns/day from 89,5 to about 135,1, that's a speedup of 50% in a real live test.

I will now run all the benchmark systems and check whether it gets slower in any of them.

I applied the following further changes in sort.cl, maybe you can confirm the correctness:

```
    if (get_global_id(0) < get_local_size(0))
        for (int index = get_local_id(0); index < length; index += get_local_size(0))
            dataBuffer[index] = data[index];
```
and
```
    if (get_global_id(0) < get_local_size(0))
        for (int index = get_local_id(0); index < length; index += get_local_size(0))
            data[index] = dataBuffer[index];
```
That avoids that data is copied repeatedly into the same locations if length > local_size.

**Comment by ThWuensche** [comment]
> Based on the CUDA profile, I _think_ this is what's happening.
> 
> We split the computation between two streams, which allows the GPU to execute two kernels at once. It turns out that sorting the list is done on one stream at the same time that `gridSpreadCharge()` is running on the other stream. That's a somewhat expensive kernel, and it's performance is entirely limited by memory access. It does a ton of writes to global memory. The short list kernel takes longer to run than the standard sorting algorithm, but it also involves a lot less memory access. That means it leaves the bus free for

As I understand the split and parallelization is done only for CUDA, right? This might actually be one of the main reasons why CUDA has improved performance over openCL.

**Comment by ThWuensche** [comment]
RUN | 7.4.2 | 7.5 limit 1024 | change(%) | min sort size
----|-------|------------------|-----|---
0 | 104.3 | 111.5 | 6.9 |
1 | 105.2 | 112.4 | 6.8 |
2 | 220.8 | 233.8 | 5.9 |
3 | 263.7 | 283.1 | 7.4 |
4 | 43.6 | 46.7 | 7.1 |
5 | 33.6 | 35.5 | 5.7 |
6 | 73.6 | 80.2 | 9.0 | 
7 | 86.6 | 92.7 | 7.0 | 
8 | 35.5 | 32.9 | -3.5 |
9 |  | | |
10 | 12.3 | 12.7 | 3.3 |
11 | 10.3 | 10.5 | 1.9 |
12 | 111.2 | 108.3 | -2.6 |
13 | 43.6 | 44.2 | 1.4 |
14 | 71.5 | 68.7 | -3.9 |
16 | 46.9 | 72.9 | 55.4 |

Columns 2 and 3 hold performance ns/day. Some cases are significantly faster with limit for sortShortList of maximum 1024, some are slightly slower. The smaller differences may result from the version difference, will check whether in that cases sortShortList would have been called at all.

**Comment by jchodera** [comment]
> As mentioned above, for the RUN9 of F@H benchmarks the limit at 1024 improves the estimated ns/day from 89,5 to about 135,1, that's a speedup of 50% in a real live test.

Wow, that's amazing!

**Comment by peastman** [comment]
> As mentioned above, for the RUN9 of F@H benchmarks the limit at 1024 improves the estimated ns/day from 89,5 to about 135,1, that's a speedup of 50% in a real live test.

But that was on AMD, right?  My numbers above were for NVIDIA.  This emphasizes the difference in the architectures, and that we'll probably need different solutions for them.

> As I understand the split and parallelization is done only for CUDA, right?

No, it's done on OpenCL as well.

**Comment by jchodera** [comment]
> But that was on AMD, right? My numbers above were for NVIDIA. This emphasizes the difference in the architectures, and that we'll probably need different solutions for them.

Could an adaptive scheme automatically select whichever algorithm is fastest for the given system so we don't hard-code this for specific GPUs?

**Comment by peastman** [comment]
Not in any easy way, especially since we have to consider the interactions between different kernels that are running at the same time.

**Comment by ThWuensche** [comment]
> No, it's done on OpenCL as well.

That might explain the reduced performance in individual cases. Interesting is that for example runs 5 and 8 have similar atom count on both lists, but performance on one is going slightly up, on the other slightly down. However difference is not big.

Really things may be complicated by the interactions between different kernels running at the same time and their ressource requirements. Anyhow the gain in performance from lower limits in the list size for shortSort for the small projects is rather significant.

sortShortList requires the lowest amount of ressources but seems to have the biggest problems with larger lists.

**Comment by jchodera** [comment]
> Not in any easy way, especially since we have to consider the interactions between different kernels that are running at the same time.

One simple approach would be to expose the choice of kernels as platform properties and allow a C++ or Python code to iterate over different options and compare timings. Even better would be a C++ API point that would allow a few different sets of options to be timed without needing to reinitialize the `Context`.

If each timestep just takes milliseconds, it should be quick to try a few timesteps with different combinations and see which one gives best performance.

**Comment by peastman** [comment]
I wanted to get more concrete data about real world performance, so I wrote the following script to generate water boxes of various sizes and time how long it takes to run a simulation.

```python
from simtk.openmm import *
from simtk.openmm.app import *
from simtk.unit import *
import math
import time

width = 3
for i in range(10):
    print('creating with width', width)
    ff = ForceField('tip3pfb.xml')
    modeller= Modeller(Topology(), []*nanometers)
    modeller.addSolvent(ff, boxSize=Vec3(width, width, width))
    system = ff.createSystem(modeller.topology, nonbondedMethod=PME)
    print('list sizes:', system.getNumParticles(), math.ceil(system.getNumParticles()/32))
    integrator = LangevinMiddleIntegrator(300, 1.0, 0.004)
    context = Context(system, integrator, Platform.getPlatformByName('OpenCL'))
    context.setPositions(modeller.positions)
    LocalEnergyMinimizer.minimize(context)
    integrator.step(10)
    context.getState(getEnergy=True)
    t1 = time.time()
    integrator.step(3000)
    context.getState(getEnergy=True)
    t2 = time.time()
    print(t2-t1, 'seconds')
    width *= 1.2
```

I ran this on the Titan V with both OpenCL and CUDA, and on the Radeon RX 5700 XT.  For each one, I ran it with the standard logic, and again with the short list kernel disabled so it would always use the standard algorithm.

On each time step it sorts two lists.  One has length equal to the number of atoms, and the other is the number of atoms divided by 32.  All the results are shown below.  Since these are times, lower is better.  Wherever the speed difference between the two versions is more than 3%, the faster one is shown in bold.

|List 1|List 2|NVIDIA OpenCL|NVIDIA OpenCL no short|CUDA|CUDA no short|AMD OpenCL|AMD OpenCL no short|
|-|-|-|-|-|-|-|-|
|2661|84|**0.77**|0.84|0.63|0.64|2.95|**2.59**|
|4527|142|1.04|1.05|0.81|**0.76**|3.89|**3.09**|
|7860|246|1.55|**1.44**|1.19|**0.98**|4.35|**2.66**|
|13695|428|**1.68**|1.77|**1.25**|1.31|3.74|**3.58**|
|23844|746|**1.61**|1.81|**1.26**|1.34|**3.84**|4.08|
|40800|1275|2.37|2.42|**1.84**|1.90|**5.16**|5.31|
|71424|2232|3.81|3.84|2.91|2.90|8.46|8.29|
|122553|3830|**6.48**|6.78|4.89|4.84|14.20|**13.74**|
|212229|6633|12.99|13.04|9.39|**9.02**|27.92|**26.87**|
|366936|11467|25.52|26.08|17.90|18.03|50.80|50.74|

The results are complicated.  And while the details vary between platforms and GPUs, the overall pattern is the same: the first rows are faster with the short list kernel disabled, then it switches over to being faster with it enabled, and then it switches back again.

Recall that the standard logic uses the short list kernel for lists shorter than 8192 elements.  In the first three rows it's using that for both lists.  In these rows, disabling the short list kernel usually makes it faster.  It then switches over to the standard sort for the longer list, but keeps using the short list kernel for the second one.  This is where it switches back to being faster with the short list kernel.  The benefit gets smaller as the lists get longer, though.  Around row 7 there is no longer much difference, and in rows 8 and 9 it goes back to being faster without the short list kernel.

In row 10, both lists are longer than 8192, so it makes no difference whether we enable the short list kernel or not.  Any differences in this row are just sampling error.

I think these are the main takeways.

1. The behavior is overall quite similar between NVIDIA and AMD, so we don't need different logic between them.
2. We should lower the size limit for the short list kernel from 8192 down to around 2000.  That will give near optimal performance in all cases.

**Comment by ThWuensche** [comment]
Here is your list with figures from Radeon VII added:

List 1 | List 2 | NVIDIA OpenCL | NVIDIA OpenCL no short | CUDA | CUDA no short | AMD OpenCL | AMD OpenCL no short | Rad VII | Rad VII no short | RadVII 1024
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
2661 | 84 | 0.77 | 0.84 | 0.63 | 0.64 | 2.95 | 2.59 | 2.27 | 1.68 | 1.62
4527 | 142 | 1.04 | 1.05 | 0.81 | 0.76 | 3.89 | 3.09 | 3.43 | 2.00 | 1.97
7860 | 246 | 1.55 | 1.44 | 1.19 | 0.98 | 4.35 | 2.66 | 4.80 | 2.24 | 2.22
13695 | 428 | 1.68 | 1.77 | 1.25 | 1.31 | 3.74 | 3.58 | 2.94 | 3.09 | 2.95
23844 | 746 | 1.61 | 1.81 | 1.26 | 1.34 | 3.84 | 4.08 | 3.26 | 3.43 | 3.27
40800 | 1275 | 2.37 | 2.42 | 1.84 | 1.90 | 5.16 | 5.31 | 4.60 | 4.57 | 4.49
71424 | 2232 | 3.81 | 3.84 | 2.91 | 2.90 | 8.46 | 8.29 | 7.77 | 7.33 | 7.31
122553 | 3830 | 6.48 | 6.78 | 4.89 | 4.84 | 14.20 | 13.74 | 12.72 | 12.04 | 11.89
212229 | 6633 | 12.99 | 13.04 | 9.39 | 9.02 | 27.92 | 26.87 | 23.05 | 21.05 | 21.03
366936 | 11467 | 25.52 | 26.08 | 17.90 | 18.03 | 50.80 | 50.74 | 36.95 | 36.86 | 36.90

While I (mostly) share your conclusion, I have considerations regarding the test procedure, I think it shows something different. At least for AMD the effect is that the larger list even from the beginning is in a range where the short algorithm is ineffective, hence the long version is faster. When list sizes are increased, we get into a region where the larger list is out of the range for short algorithm, but the smaller list is still in a region where the short algorithm is effective, hence the switchover. Then also the smaller list gets ineffective with the short algorithm - switch back. In the end both are beyond the limit, equal results. For NVidia the situation is slightly different, since there the short algorithm does not loose efficiency so quick, raise of times is slower.

The last column is with switchover limit set to 1024. I think, for AMD the limit might be lower, somewhere around 1000. But 2000 probably is still acceptable. If possible, I would suggest 2000 for NVidia, as you suggest, but 1000 for AMD.

**Comment by ThWuensche** [comment]
Using this as scratchpad, I still don't understand the issue with tilesize and wavefront size. I somehow assumed that the tilesize also is the wavefront size and that way the tilesize would limit the use of the wavefront to 1/2 on GCN CUs. I saw that in the HIP port the mask was extended to long, thus allowing tilesize 64. But then if on GCN CUs only 1/2 of the wavefront size were used, performance should be lower. So probably I do not understand the relation between tilesize and wavefront size. What happens if tilesize is 32, but wavefront size is 64? Are two tiles executed in one wavefront?

**Comment by peastman** [comment]
> Using this as scratchpad, I still don't understand the issue with tilesize and wavefront size.

They're mostly unrelated to each other.

We divide atoms into blocks of 32.  We then compute nonbonded interactions in 32x32 tiles, that is, all the atoms in one block interacting with all the atoms in the other block.

The reason for doing this is to reduce memory access.  In order to compute an interaction between two atoms, you need to load the parameters for both atoms.  If we treated every interaction as independent, we would have to load parameters for two atoms every time we computed one interaction.  That is, a 2:1 ratio of atoms loaded to interactions computed.  Instead we load the data for 64 atoms (two blocks of 32) and compute a tile of 1024 interactions.  That's a 1:16 ratio of atoms loaded to interactions computed.

The larger the tile size, the better that ratio gets.  But it also creates a problem: it sets the granularity of our neighbor list.  If an atom is within the cutoff distance of *any* atom in a block, then we have to pay the cost of computing its interaction with *every* atom in the block.  The result is that, as you increase the tile size, the memory overhead goes down but the compute overhead goes up.  32 strikes a good balance between these two effects.

**Comment by ThWuensche** [comment]
Ok, thanks, now I see more clear. Actually there is a two-dimensional matrix (table) with the sides (like row header, column header) being defined by the tiles and the atoms within them (like one tile used as a row and one as a column vector), the elements of the matrix (table) being the interaction between the two atoms at the (row, column) header at the side of the table. Computation (wavefront) is unrelated to that and created as needed by the equations to calculate the matrix elements.

So - as you mentioned before, but I did not understand :) - no reason to change tile size for purposes of wavefront construction.

**Comment by peastman** [comment]
> If possible, I would suggest 2000 for NVidia, as you suggest, but 1000 for AMD. 

Possibly it would be best to use different cutoffs for them.  I've been running more closely spaced tests to narrow down the crossover points:

```python
width = 8
...
  width += 0.25
```

On Titan V, using the short list kernel is consistently faster when the list length is below 2000.  Between 2000 and 3000 the difference between the two versions is minimal.  Can you try the same on Radeon VII and see if you can narrow down where the crossover happens?

**Comment by ThWuensche** [comment]
width 8 generates to big lists already in the starting point, something like 1577 for the smaller list. Running with width 4, is that ok?

Or could we have a setup where only one list changes size?

**Comment by peastman** [comment]
For AMD it looks like the transition will probably happen at a width somewhere between 7 and 9.

**Comment by ThWuensche** [comment]
Seems to be around 1000 for Radeon VII

List 1 | List 2 | Std | NoShort
-- | -- | -- | --
12255 | 383 | 2.79 | 2.93
14190 | 444 | 2.96 | 2.99
16284 | 509 | 2.67 | 2.80
18600 | 582 | 2.94 | 3.01
21384 | 669 | 3.13 | 3.22
24144 | 755 | 3.31 | 3.37
26964 | 843 | 3.49 | 3.53
30075 | 940 | 3.70 | 3.77
33840 | 1058 | 4.13 | 4.08
37428 | 1170 | 4.42 | 4.32

**Comment by peastman** [comment]
Could you continue it up to slightly higher lengths?  The noise level in these measurements is usually around 3%.  The numbers at the bottom of the table are still close enough that they may not reflect real differences.

**Comment by ThWuensche** [comment]
I have extended the table, please see above. Times really are not very predictable, see the step backward in time consumption from 2nd to 3rd line. Applies for both cases, so probably lies within the standard (long) algorithm (or some other calculations). Do you have an explanation, or might that hide another issue? How many threads does that algorithm use, is that fixed or does it rise with size of the list? You mentioned it needs 5 kernels, but how many threads do they occupy? Also interesting is the sharp increase after the line with 45792/1413.

**Comment by peastman** [comment]
> Applies for both cases, so probably lies within the standard (long) algorithm (or some other calculations).

No idea what the reason is!  It's probably not related.  Changing the system size affects every part of the calculation, and sorting is only a small piece.  The important thing is the difference in times between using the two sorting algorithms.

Based on your data, here are the ratios of the times using the two algorithms as a function of list length.

|Length|Ratio|
|-|-|
|383|0.95221843003413|
|444|0.989966555183946|
|509|0.953571428571429|
|582|0.976744186046512|
|669|0.972049689440994|
|755|0.982195845697329|
|843|0.988668555240793|
|940|0.981432360742705|
|1058|1.01225490196078|
|1170|1.02314814814815|
|1293|1.03418803418803|
|1431|1.03012048192771|
|1577|1.02337228714524|
|1730|1.01401869158879|
|1889|1.04097116843703|

We see two things: first that the crossover point is probably somewhere between 940 and 1058, and second that the differences are very small over most of this range.  1024 is probably as good a place as any to do the switch on AMD.  On NVIDIA we want it to be a bit higher.  I make the change today.

**Comment by ThWuensche** [comment]
Good!

The most critical cases before have been with list sizes beyond 4000, that's not visible here, but in that [table](https://github.com/openmm/openmm/issues/2875#issuecomment-704552767) row 3 with 4.8s/2,2s. The fact that differences are small probably results from the fact that in this system effort for sort is only a very small part of the whole computation effort and maybe shaded by other calculations in parallel. The systems where I had seen significant influence had groups of 128 and 4071 atoms, there the larger list is in the critical range and probably other calculations are not that much.

In that small system according to the trace only about 2/3 of wall time are occupied by kernel time, at minimum one third of the time the GPU seems to be idle. So kernel schedule latency has a big effect. That's another point I just try to evaluate, but probably more related to the underlying system.

**Comment by peastman** [comment]
The change is in #2878.

**Comment by informatorius** [comment]
@ThWuensche: You can try to run 2 instances of performance benchmarks in parallel. If both take double the time then there were no idle resources left on GPU. If e.g. both complete in same time then GPU was half idle with only one simulation running.

**Comment by peastman** [comment]
It doesn't always work that way.  GPUs aren't nearly as good as CPUs at multitasking.
